### PR TITLE
Optimize for size and performance the most common log message

### PR DESCRIPTION
**Pull request by nordic-krch**
It's been noticed that logging messages takes more ROM on RISCV platform others (like ARM Cortex-M). On RISCV32 message takes 70-80 bytes and on Cortex-M 30-40 bytes. This PR is an attempt to improve that by focusing on the most common log messages and optimizing them and using legacy generic approach whenever optimized approach cannot be applied.

I've analyzed zephyr code base (and other projects) which log messages are the most common and got following results:
- 45% of all logs are strings without arguments (e.g. `LOG_INF("foo")`)
- 24% of all logs are strings with one 32 bit argument (e.g.` LOG_INF("char: %c", c)`)
- 6 % of all logs are strings with two 32 bit arguments (e.g. `LOG_INF("%08x %d, a, b)`)

Overall there is usually between 75%-80% of those "simple" messages in the application. On 32 bit platforms messages like that can be built very easily since there is no alignment or padding to consider. Based on that analysis, targeted approach was taken to provide optimized macros and functions for logging "simple" messages on 32 bit architectures.

Optimization is enabled by default (`CONFIG_LOG_SIMPLE_MSG_OPTIMIZE=n` to disable).

Optimization results on RISCV32 (~40% less code)

| Log type    | Before  | After |
| -------------- | ---------- | ---- 
| Log 0 args | 46     | 30 |
| Log 1 arg   | 68     | 38 |
| Log 2 args | 70     | 40 |

Optimization results on ARM Cortex-M3 (~10% less code)

| Log type    | Before  | After |
| -------------- | ---------- | ---- 
| Log 0 args | 36     | 28 |
| Log 1 arg   | 44     | 32 |
| Log 2 args | 44     | 40 |

If we consider RISCV32 application with 300 log messages with average distribution of message types that means 5kB code saving. Improvements on ARM Cortex-M3 are not that significant but they are seen in performance where "simple" logs logging went from **14us** to **10us**. On RISCV32 performance gain is approx.40%.

PR extends also log_frontend API with optional functions for handling "simple" messages.

### CODE REVIEW

File: include/zephyr/logging/log_frontend.h [hunk 1]
```diff
@@ -32,6 +32,52 @@ void log_frontend_msg(const void *source,
...
+/** @brief Log message with 0 arguments.
+ *
+ * Optimized version for log message which does not have arguments (only string).
+ * This API is optional and is used only if optimizing common log messages is enabled.
+ *
+ * @param source Pointer to a structure associated with given source. It points to
+ * static structure or dynamic structure if runtime filtering is enabled.
+ * @ref log_const_source_id or @ref log_dynamic_source_id can be used to determine
+ * source id.
+ * @param desc Message descriptor.
```
Comment (pdgendt) [inline include/zephyr/logging/log_frontend.h:44]: The parameter `desc` does not exist?

File: include/zephyr/logging/log_core.h [hunk 2]
```diff
@@ -290,7 +290,7 @@ static inline char z_log_minimal_level_to_char(int level)
 		break; \
 	} \
 	/* For instance logging check instance specific static level */ \
-	if (_inst & !IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING)) { \
+	if (_inst && !IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING)) { \
```
Comment (pdgendt) [inline include/zephyr/logging/log_core.h:293]: Was this a bug? Maybe move this to a new PR?

File: tests/drivers/counter/counter_basic_api/src/test_counter.c [hunk 3]
```diff
@@ -921,9 +923,19 @@ static void test_cancelled_alarm_does_not_expire_instance(const struct device *d
 
 		cnt = IS_ENABLED(CONFIG_ZERO_LATENCY_IRQS) ?
 			alarm_cnt : k_sem_count_get(&alarm_cnt_sem);
-		zassert_equal(0, cnt,
+		/* For slower processors it may happen that alarm expires before
+		 * it is canceled. That is accepted scenario and it is detected
+		 * by checking that timestamp in the last alarm handler was for
+		 * the current alarm.
+		 */
+		zassert_true((cnt == 0) || (last_alarm_handler >= alarm_cfg.ticks),
```
Comment (pdgendt) [inline tests/drivers/counter/counter_basic_api/src/test_counter.c:931]: Unrelated changes? Move to another PR?

Comment (nordic-krch) [reply tests/drivers/counter/counter_basic_api/src/test_counter.c:931]: Ups, Started from wrong branch. Will remove.

File: include/zephyr/logging/log_core.h (hunk 2 above)
Comment (nordic-krch) [reply include/zephyr/logging/log_core.h:293]: Good point. Opened #64170. I will remove this commit from that PR.

File: include/zephyr/sys/cbprintf_internal.h [hunk 4]
```diff
@@ -113,6 +113,31 @@ extern "C" {
...
+/** @brief Check if argument fits in 32 bit word.
+ *
+ * @param x Input argument.
+ *
+ * @retval 1 if variable is of type that fits in 32 bit word.
+ * @retval 0 if variable is of different type.
+ */
+#ifdef __cplusplus
+#define Z_CBPRINTF_IS_WORD_NUM(x) \
+	z_cbprintf_cxx_is_word_num(x)
```
Comment (cfriedt) [inline include/zephyr/sys/cbprintf_internal.h:125]: The irony that C++ compilers cannot handle C11 generics... 🙄

**Review by cfriedt — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]
